design-proposals/klip-58-limit-clause-for-pull-queries.md

```diff
@@ -0,0 +1,217 @@
+steps of the physical plan till it is full and then return.
+
+As an example, if we have a table called `ridersNearMountainView` with 
+`Schema: profileId VARCHAR, latitude DOUBLE, longitude DOUBLE`, and we execute a pull query of the form:
+
+```roomsql
+SELECT profileid, latitude 
+FROM riderlocations 
+WHERE latitude > 38 
+LIMIT 5;
+```     
+then the query gets translated into a logical plan which in turn gets translated into a physical plan that look like:
+```
+     Logical Plan                     Physical Plan
+    ──────────────                   ───────────────
+                              `TransientQueryMetadata` has 
+                      ┌────────►    limit information 
+                      │
+  ┌────────────────┐  │              ┌──────────────┐
+  │ DataSourceNode │  │              │ StreamSource │
+  └────────┬───────┘  │              └───────┬──────┘
+           │          │                      │       
+    ┌──────▼─────┐    │              ┌───────▼──────┐
+    │ FilterNode │    │              │ StreamFilter │
+    └──────┬─────┘    │              └───────┬──────┘
+           │          │                      │       
+  ┌────────▼───────┐  │                      │
+  │ QueryLimitNode │──┘    ────────►         │
+  └────────┬───────┘                         │
+           │                                 │
+ ┌─────────▼────────┐                ┌───────▼──────┐
+ │ FinalProjectNode │                │ StreamSelect │
+ └─────────┬────────┘                └──────────────┘
+           │
+┌──────────▼─────────┐
+│ KsqlBareOutputNode │
+└────────────────────┘
+```
+The output of the above query will look like:
+```
++-------------------------------+-------------------------------+
+|PROFILEID                      |LATITUDE                       |
++-------------------------------+-------------------------------+
+|c2309eeca                      |38.7877                        |
+|18f4ea86a                      |38.3903                        |
+|4ab5cbada                      |38.3952                        |
+|8b6eae59a                      |38.3944                        |
+|4a7c7b41a                      |38.4049                        |
+Limit Reached
+Query terminated
+```
+
+If the limit isn't reached and we run out stream records to return (exhausting the records that conform to our query), 
+then the `TransientQueryQueue` will just be returned. 
+
+## Test plan
+- Unit testing.
+- Functional testing to ensure that the queries generate the correct physical and logical plans for pull queries on 
+streams and tables.
+- Rest query validation tests to ensure that we are returning the correct number of rows and ending the query with 
+messages saying `Limit Reached` and `Query terminated` for pull queries on streams and tables.
+- Queries with limit clause argument that is not a positive integer should throw a ksqlDB syntax exception and output an 
```

Comment:
Let's support `LIMIT 0` as it can be useful for checking the validity of a query. It can also be employed to obtain the types of the result columns quickly.